### PR TITLE
Introduce the 1st Gen 64 and 128 Heads MLA Decode Kernel for DeepSeek V4 for MI35x

### PR DESCRIPTION
**Pull request by ruanjm**
## Introduction

This PR lands the **Gen.1** implementation of the V40 MLA decode kernel (`mi35x_v40_fwd_decode_m16x8_fp8bf16_fp8bf16_gen1`) for AMD MI350 (gfx950). It targets the DeepSeek-style V4 MLA layout (FP8 NoPE 448 + BF16 RoPE 64, $D_V = 512$) and supports any $(H, \mathrm{mtp})$ packing with $H \cdot \mathrm{mtp} = 64\ or\ 128$.

This Gen.1 design is **relatively conservative**: every input is converted to **bf16 before** the GEMMs, so QK and PV both run on `v_mfma_f32_16x16x32_f16` rather than directly consuming fp8. KV cvt+scale happens on the LDS-store path; Q cvt+scale happens once at the prologue. This trades a fraction of peak fp8 mfma throughput for simpler scheduling and a wider compatibility surface.

A **Gen.2** rewrite is planned that will keep more of the data in its native fp8 form through QK, double-buffer the V load on the PV side (the `pv_v_aux` slot reserved but unused in Gen.1), and switch to an m32×4 partition (4 ptiles per workgroup, 2 waves per ptile). The expected outcome is a meaningful step up in long-context decode throughput vs the
numbers reported below.

The full design spec is in
`csrc/kernels/mla/hk/doc/hk_mla_v40_gen1_spec.md`.

---

## Setup

- **GPU**: AMD MI350 (gfx950), `HIP_VISIBLE_DEVICES=4` (clean, idle GPU)
- **Branch**: `jruan/hk_mla_v4`
- **Harness**: `op_tests/test_mla_v4_persistent.py` (`AITER_ENABLE_EXPERIMENTAL=1`)
- **Tolerance**: `atol = rtol = 0.01` (silver vs aiter_v40)
- **Build**: clean JIT rebuild of `module_hk_mla_v40_fwd` before measurements
- **HIP version:** 7.2.53211-671d39a71e
- **AMD clang version:** 22.0.0git (https://github.com/RadeonOpenCompute/llvm-project roc-7.2.2 26084 f58b06dce1f9c15707c5f808fd002e18c2accf7e)

## Quality (silver vs aiter_v40)

Cross-product of `page_size ∈ {1, 64}` × all m16x8-valid $(H, \mathrm{mtp})$ combos, on `-b 4 -c 4096`:

| $(H, \mathrm{mtp})$ | page_size = 1 | page_size = 64 |
|---|:---:|:---:|
| 128, 1 | PASS | PASS |
| 64, 2  | PASS | PASS |
| 32, 4  | PASS | PASS |
| 16, 8  | PASS | PASS |

Additional long-context sanity check (`b=33, c=63333, n=128,1`):

| page_size | silver |
|---:|:---:|
| 1  | PASS |
| 64 | PASS |

V32 regression check (after the manager split that moved shared helpers into `namespace hk_mla`):

| | Result |
|---|:---:|
| `module_hk_mla_v32_fwd` JIT build | PASS |
| `aiter.hk_mla_v32_decode_fwd` symbol resolves | PASS |

## Performance (v1.1)

  Measured on `520799732`, decode kernel only (the `mla_reduce_v1` pass is excluded, matching the v1 numbers). Same toolchain as v1 (HIP `7.2.53211-671d39a71e`, AMD clang `22.0.0git roc-7.2.2 26084`).
  `HIP_VISIBLE_DEVICES=0`, `AITER_ENABLE_EXPERIMENTAL=1`, `op_tests/test_mla_v40_persistent.py`.

  v1.1 adds the **m16x4** partition (`H·mtp = 64`), new since v1.

### Short context (`b=4, c=4096`)

  **m16x8 (`H·mtp = 128`)**

  | $(H, \mathrm{mtp})$ | page_size | µs | TFLOPS | TB/s |
  |---|---:|---:|---:|---:|
  | 128, 1 | 1  |  23.99 | 179.01 | 0.432 |
  | 64, 2  | 1  |  23.83 | 180.24 | 0.435 |
  | 32, 4  | 1  |  24.13 | 177.96 | 0.430 |
  | 16, 8  | 1  |  24.01 | 178.86 | 0.432 |
  | 128, 1 | 64 | 194.27 |  22.11 | 0.0534 |
  | 64, 2  | 64 | 194.65 |  22.06 | 0.0533 |
  | 32, 4  | 64 | 194.63 |  22.07 | 0.0533 |
  | 16, 8  | 64 | 194.47 |  22.09 | 0.0533 |

  **m16x4 (`H·mtp = 64`, new in v1.1)**

  | $(H, \mathrm{mtp})$ | page_size | µs | TFLOPS | TB/s |
  |---|---:|---:|---:|---:|
  | 64, 1 | 1  | 17.85 | 120.29 | 0.558 |
  | 32, 2 | 1  | 17.97 | 119.49 | 0.554 |
  | 16, 4 | 1  | 17.83 | 120.42 | 0.559 |
  | 64, 1 | 64 | 161.57 | 13.29 | 0.0617 |
  | 32, 2 | 64 | 161.44 | 13.30 | 0.0617 |
  | 16, 4 | 64 | 161.58 | 13.29 | 0.0617 |

  (`(8, 8)` is rejected by the metadata planner's supported-shape set, so m16x4 is measured at `(64,1)/(32,2)/(16,4)`.)

### Long context (`b=33, c=63333, n=H,1`)

  **m16x8 (`128, 1`)**

  | page_size | µs | TFLOPS | TB/s |
  |---:|---:|---:|---:|
  | 1  | 499.55 | 1096.74 | 2.45 |
  | 64 | 615.32 |  890.40 | 1.99 |

  **m16x4 (`64, 1`, new in v1.1)**

  | page_size | µs | TFLOPS | TB/s |
  |---:|---:|---:|---:|
  | 1  | 328.63 | 833.57 | 3.72 |
  | 64 | 442.97 | 618.42 | 2.76 |

### v1 → v1.1 (m16x8, decode-only, same methodology)

  | workload | v1 µs | v1.1 µs | speedup |
  |---|---:|---:|---:|
  | short `b4 c4096` (128,1), pg 1  |  32.46 |  23.99 | 1.35× |
  | short `b4 c4096` (128,1), pg 64 | 269.80 | 194.27 | 1.39× |
  | long `b33 c63333` (128,1), pg 1  | 633.92 | 499.55 | 1.27× |
  | long `b33 c63333` (128,1), pg 64 | 806.24 | 615.32 | 1.31× |

  `page_size = 64` remains slower than `page_size = 1` at the same workload — same ISA path, the gap is the metadata planner's work distribution under page-size 64 (unchanged from v1).

## HK MLA v4.0 Decode vs. asm `v4_nm`: Performance Comparison

  Both kernels implement the same DeepSeek-V4 MLA decode (FP8 NoPE + BF16 RoPE, 512-dim, page_size=1). This compares the HK persistent kernel (`mla_v40_decode_fwd`) against the shipped asm kernel (`mla_decode_fwd_v4_nm`) on gfx950 (MI350, 256 CUs), and explains the difference at the hardware level.

  ### TL;DR

  - **At scale (large ctx / high batch / splits on), HK is ~2x faster.** The cause is occupancy: HK runs **2 waves/CU**, v4_nm is hard-capped at **1**.
  - **At low load (single split, small work), the gap shrinks to ~1.2x** — pure per-wave compute efficiency, since there is no second wave to schedule.
  - The earlier impression that "v4_nm wins" was an artifact of split-heuristic tuning, **not** raw kernel efficiency. With both tuned to fill the CUs, HK's decode kernel is faster at every context length.

  ### 1. Fair no-split comparison (no KV split either side, matched ctx)

  Isolates raw kernel work. v4_nm = asm decode only; HK broken into decode kernel vs. the reduce it always pays. Times in us, gfx950, idle GPU, batch=1.

  | ctx    | v4_nm | HK decode | HK reduce | HK total | winner (decode) |
  |-------:|------:|----------:|----------:|---------:|:----------------|
  | 256    | 19.7  | 16.3      | 1.8       | 18.0     | HK 1.21x        |
  | 1024   | 54.1  | 44.0      | 1.7       | 45.7     | HK 1.23x        |
  | 4096   | 191.5 | 154.9     | 1.7       | 156.6    | HK 1.24x        |
  | 16384  | 745.0 | 600.2     | 1.7       | 602.0    | HK 1.24x        |
  | 65536  | 2953  | 2374      | 1.8       | 2376     | HK 1.24x        |

  HK's decode kernel is ~1.2x faster than v4_nm at every ctx (nh=64). HK's reduce is a flat ~1.7-2.3 us; it only matters as a fraction at the smallest ctx.

 ### 2. Occupancy is the ceiling (root cause of the ~2x at scale)

  | Resource            | v4_nm                     | HK           |
  |---------------------|---------------------------|--------------|
  | arch VGPR / lane    | 256                       | 256          |
  | AGPR / lane         | **256**| **0**        |
  | Total register file | **512 (full)**            | 256          |
  | LDS / block         | **160 KB (all of it)**    | 80 KB        |
  | **Occupancy (measured)** | **1.00 wave/SIMD**     | **2.00 waves/SIMD** |

  v4_nm holds its accumulator tile in 256 AGPRs and grabs all 160 KB of LDS — either constraint alone pins it at 1 resident wave/CU. HK's lean footprint (0 AGPR, 80 KB LDS) leaves room for 2. Measured via `MeanOccupancyPerActiveCU` at batch = 2xCU (512 workgroups, no split): v4_nm 1.00, HK 1.99.

  ### 3. Thread trace: HK hides stalls, v4_nm exposes them

  ATT on the target CU (b=512, ctx=4096):

  | Metric               | v4_nm (1 wave) | HK (2 waves) |
  |----------------------|---------------:|-------------:|
  | Stall % of cycles    | 26.6%          | 37.1%        |
  | Idle % of cycles     | 4.2%           | 3.0%         |
  | MFMA stall cycles    | 92K            | 62K          |

  HK has *higher* per-wave stall% yet is faster and has *lower* idle: when one wave stalls on `s_waitcnt` (LDS/VMEM), the CU issues from the second resident wave. v4_nm at 1 wave/CU has nothing to overlap its stalls with, so they run exposed as dead CU time. HK also stalls less on MFMA (matches higher MfmaUtil, 32.6% vs 27.4%).

  ### 4. Why the advantage is size-dependent

  - **Enough work to fill 2 waves** (large ctx / high batch / splits): HK's second wave hides latency -> ~2x.
  - **Not enough work** (single split, small batch): both sit at ~1 wave/CU, HK's second slot is empty -> gap collapses to the ~1.2x per-wave compute-efficiency edge (better MFMA feeding, ~45x fewer LDS bank conflicts).

**Special thanks to @kaiyang-1 for his help with version V1.1.**

### CODE REVIEW

File: aiter/mla.py [hunk 1]
```diff
@@ -637,6 +637,141 @@ def mla_decode_fwd(
...
+
+    aiter.mla_reduce_v1(
+        logits,
+        attn_lse,
+        reduce_indptr,
+        reduce_final_map,
+        reduce_partial_map,
+        max_seqlen_q,
+        0,
+        o,
```
[the remark this answers is not in the document]
Comment (ruanjm) [reply aiter/mla.py:766]: 0 is `auto`. Added comments for `mla_reduce_v1`

**Review by amd-ruitang3 — APPROVED** [review]
